examples/raft-kv-rocksdb/src/store.rs

```diff
@@ -0,0 +1,564 @@
+use rocksdb::DB;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::ExampleNodeId;
+use crate::ExampleTypeConfig;
+
+/**
+ * Here you will set the types of request that will interact with the raft nodes.
+ * For example the `Set` will be used to write data (key and value) to the raft database.
+ * The `AddNode` will append a new node to the current existing shared list of nodes.
+ * You will want to add any request that can write data in all nodes here.
+ */
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum ExampleRequest {
+    Set { key: String, value: String },
+}
+
+/**
+ * Here you will defined what type of answer you expect from reading the data of a node.
+ * In this example it will return a optional value from a given key in
+ * the `ExampleRequest.Set`.
+ *
+ * TODO: SHould we explain how to create multiple `AppDataResponse`?
+ *
+ */
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ExampleResponse {
+    pub value: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ExampleSnapshot {
+    pub meta: SnapshotMeta<ExampleNodeId>,
+
+    /// The data of the state machine at the time of this snapshot.
+    pub data: Vec<u8>,
+}
+
+/**
+ * Here defines a state machine of the raft, this state represents a copy of the data
+ * between each node. Note that we are using `serde` to serialize the `data`, which has
+ * a implementation to be serialized. Note that for this test we set both the key and
+ * value as String, but you could set any type of value that has the serialization impl.
+ */
+#[derive(Serialize, Deserialize, Debug, Default, Clone)]
+pub struct SerializableExampleStateMachine {
+    pub last_applied_log: Option<LogId<ExampleNodeId>>,
+
+    // TODO: it should not be Option.
+    pub last_membership: EffectiveMembership<ExampleNodeId>,
+
+    /// Application data.
+    pub data: BTreeMap<String, String>,
+}
+
+impl From<&ExampleStateMachine> for SerializableExampleStateMachine {
+    fn from(state: &ExampleStateMachine) -> Self {
+        let mut data = BTreeMap::new();
+        for (key, value) in state.data.iterator_cf(
+            state.data.cf_handle("data").expect("cf_handle"),
+            rocksdb::IteratorMode::Start,
+        ) {
+            let key: &[u8] = &key;
+            let value: &[u8] = &value;
+            data.insert(
+                String::from_utf8(key.to_vec()).expect("invalid key"),
+                String::from_utf8(value.to_vec()).expect("invalid data"),
+            );
+        }
+        Self {
+            last_applied_log: state.last_applied_log.clone(),
+            last_membership: state.last_membership.clone(),
+            data,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ExampleStateMachine {
+    pub last_applied_log: Option<LogId<ExampleNodeId>>,
+
+    // TODO: it should not be Option.
+    pub last_membership: EffectiveMembership<ExampleNodeId>,
+
+    /// Application data.
+    pub data: Arc<rocksdb::DB>,
+}
```

Comment:
It looks like openraft conflicts with the non-persisted field `last_applied_log` and `last_membership`:

Openraft assumes once a log entry is applied with `apply_to_state_machine()`, the result is persisted. There is no need to rebuild the state machine when a node restarts.

In this example, when a node is restarted, it depends on the snapshot to rebuild the state machine, in `ExampleStore::new()`. Since the application data is written to `data: Arc<rocksdb::DB>`, it may be already persisted when `apply_to_state_machine()`. Hence rebuilding the state machine from a snapshot may re-apply an entry for the second time. This will lead to an inconsistent state, although in this example there won't be such a problem because there is only one `set` operation, which is idempotent and is safe to apply more than one time.

A state machine has to be internally consistent: All its content is persisted when `apply_to_state_machine()` is called, or none of its content is persisted.
In the latter case, it will have to rebuild from a snapshot as it does in this example.

But, to implement an app in the second way, openraft still needs an improvement:
Currently, it does not wait for a `build_snapshot()` to finish before deleting applied log entries. This means there may not be an up-to-date snapshot to rebuild the state machine. To fix this, openraft has not to delete any logs that are not included in the latest snapshot.

Reply:
That makes a lot of sense! Thanks for the context I'll remove the snapshot rebuildy part :), I suspect that last_membership and last_applied_log need to be persisted in rocks then too, I'll update those with the removal